common/include/opae/metrics.h

```diff
@@ -0,0 +1,113 @@
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/**
+* @file metrics.h
+* @brief Functions for Discover/ Enumerates metrics and retrieves values
+*
+*
+*
+*
+*/
+
+#ifndef __FPGA_METRICS_H__
+#define __FPGA_METRICS_H__
+
+#include <opae/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Enumerates number of metrics
+ *
+ * @param[in]  handle     Handle to previously opened fpga resource
+ * @param[inout] num_metrics number of metrics are discovered in
+ *                        fpga resource
+ * @returns FPGA_OK on success. FPGA_NOT_FOUND if the Metrics are not
+ * discovered
+ *
+ */
+fpga_result fpgaGetNumMetrics(fpga_handle handle,
+				uint64_t *num_metrics);
+
+/**
+ * Retrieve metrics information
+ *
+ * @param[in]  handle     Handle to previously opened fpga resource
+ * @param[inout] metric_info pointer to array of metric info struct
+ *                        user allocates metrics info array
+ * @param[inout] num_metrics size of metric info array
+ *
+ * @returns FPGA_OK on success. FPGA_NOT_FOUND if the Metrics are not
+ * found
+ *
+ */
+fpga_result fpgaGetMetricsInfo(fpga_handle handle,
+				fpga_metric_info *metric_info,
+				uint64_t *num_metrics);
+
+/**
+ * Retrieve metrics values by index
+ *
+ * @param[in]  handle     Handle to previously opened fpga resource
+ * @param[inout] metric_num pointer to array of metric index
+ *                        user allocates metric array
+ * @param[inout] num_metric_indexs size of metric array
+ * @param[inout] metrics pointer to array of metric struct
+ * @returns FPGA_OK on success. FPGA_NOT_FOUND if the Metrics are not
+ * found
+ *
+ */
+fpga_result fpgaGetMetricsByIndex(fpga_handle handle,
+				uint64_t *metric_num,
+				uint64_t num_metric_indexes,
+				fpga_metric *metrics);
+
+/**
+ * Retrieve metric values by names
+ *
+ * @param[in]  handle     Handle to previously opened fpga resource
+ * @param[inout] metrics_names pointer to array of metrics name
+ *                        user allocates metrics name array
+ * @param[inout] num_metric_names size of metric name array
+ * @param[inout] metrics pointer to array of metric struct
+ * @returns FPGA_OK on success. FPGA_NOT_FOUND if the Metrics are not
+ * found
```

Comment:
are there any other possible return codes? Is it possible that some code path leads to something returning FPGA_NO_MEMORY?